Fix listing collections that are missing the metadata required by build

s-hertel opened the pull request:
##### SUMMARY
Fixes #70180
Refreshed version of #70185

Don't error while listing collections if a collection's metadata is missing keys required for building a collection.

Give an informative error if metadata has been badly formatted.

As mentioned in #70185, threading this through the existing functions just for `list` doesn't seem ideal. An alternative would be creating a new function just for list rather than using the common methods, but it duplicates code and could allow inadvertent divergence.

##### ISSUE TYPE
- Bugfix Pull Request

webknjaz commented on lib/ansible/cli/galaxy.py:
```diff
@@ -1485,7 +1485,7 @@ def execute_list_collection(self, artifacts_manager=None):
                     display.vvv("Searching {0} for collections".format(collection_path))
                     collections = list(find_existing_collections(
-                        collection_path, artifacts_manager,
+                        collection_path, artifacts_manager, require_mandatory_keys=False
```
Mind keeping the trailing comma?
```suggestion
                        collection_path, artifacts_manager, require_mandatory_keys=False,
```

s-hertel replied:
Rather than thread this through the entirety of find_existing_collections, I thought it might be easier to read if I hang it on the ConcreteArtifactManager. Now it's a property that execute_list_collection sets.

webknjaz commented on changelogs/fragments/70180-collection-list-more-robust.yml:
```diff
@@ -0,0 +1,2 @@
+bugfixes:
+  - ansible-galaxy - do not require mandatory keys in ``galaxy.yml`` for a collection (https://github.com/ansible/ansible/issues/70180)
```
any collection?

s-hertel replied:
whoops

s-hertel replied:
Changelog now mentions the important parts - applies to source collections when listing collections.

webknjaz commented on lib/ansible/galaxy/collection/concrete_artifact_manager.py:
```diff
@@ -474,7 +476,7 @@ def _normalize_galaxy_yml_manifest(
     missing_keys = mandatory_keys.difference(set_keys)
-    if missing_keys:
+    if missing_keys and require_mandatory_keys:
         raise AnsibleError("The collection galaxy.yml at '%s' is missing the following mandatory keys: %s"
```
What if instead of ignoring the problems we'd accumulate them in a list and implement a postponed console output + erroring out?

s-hertel replied:
That seems like a good idea. I'll take a stab at that tomorrow.

bcoca replied:
you can output warnings immediately (they dedupe) and that should not interrupt the other output that happens later

s-hertel replied:
Yeah, I think just warning is the best solution. Trying to shuffle around the errors to accumulate them at the end got pretty messy. Now there is a warning for galaxy.yml's that aren't a MutableMappings and for missing keys required for building.

My patch looks strange because it's displaying a warning and then raising a ValueError with the same error, but we silently ignore the ValueErrors (in `_ComputedReqKindsMixin.from_dir_path_as_unknown`), and use them to fall back to the path-derived metadata. Adding warnings for ValueErrors in `_ComputedReqKindsMixin.from_dir_path_as_unknown` would make sense to me but then `_ComputedReqKindsMixin.from_dir_path` and maybe other things need to be refactored to not display extra warnings. So this is maybe something to refactor in the future.

s-hertel committed:
Rethread pr/70185 through the dependency resolver

Hang optional metadata toggle on the ConcreteArtifactsManager instead of threading it through whole list codepath

Don't error while listing collections if a collection's metadata is missing keys required for building a collection.

Give an informative warning if metadata has been badly formatted.

Co-authored-by: Sam Doran <sdoran@redhat.com>

s-hertel committed:
pep8

s-hertel committed:
Fix mypy sanity test. Verify it's a dict, not MutableMapping.

bcoca commented:
/azp run
